scripts/common/primer_autodetection.py

```diff
@@ -79,52 +78,44 @@ def build_primers_automaton(primer_index: Path) -> dict[str, PrimerAutomaton]:
     """
     Perform an exact search of the primer sequences (all schemes) in the sample_fastq
     """
     # Structure for storing the unique primer sequences
     unique_hits: dict[str, set] = {p: set() for p in primers_automaton}
 
+    n_reads = 0
+
     # look up the primer sequences in the automaton
     # reading the sample fastq once for all
     with gzip.open(sample_fastq, "rt") as fastq_file:
         # for each sample read, it counts the number of primers found for each primer scheme.
         # This search is performed using an ahocorasick structure and optimised to search
         # at the beginning of the read only
         for sample_sequence in SeqIO.parse(fastq_file, "fastq"):
-            sample_read = str(sample_sequence.seq)
+            # fastq_file is assumed to be adapter-trimmed, e.g.
+            # * trim_galore (illumina)
+            # * porechop (ont)
+            # Trimmer algorithms might not remove the adapter sequences entirely.
+            # Therefore, here the primer is searched in a prefix window of primer_base_window bases,
+            # rather than at the very beginning of the read.
+            sample_read = str(sample_sequence.seq)[0:primer_base_window]
+
+            n_reads += 1
 
             for primer in primers_automaton:
                 primer_auto = primers_automaton[primer]
-                # search for primers matching the prefix of sample_read
-                # (the wildcard argument is mandatory, but unused - no base will ever be '?'), e.g.
-                # keys: ['AGA', 'AAGA', 'ATGA', 'GTAT']
-                # sample_read: 'AAGATT'
-                # output: ['AAGA']
-                found_primer = set(primer_auto.automaton.keys(sample_read, "?", ahocorasick.MATCH_AT_MOST_PREFIX))
-                # search for primers matching the prefix of sample_read (wildcard 'N' is used).
-                # This enables the counting of primers upon uncertainty in sample reads. e.g.
-                # keys: ['AGA', 'AAGA', 'ATGA', 'GTAT']
-                # sample_read: 'ANGATT'
-                # output: ['AAGA', 'ATGA']
-                # Depending on the number of Ns in sample_read, this can return:
-                # - "sample_read does not contain N in primer sequence"
-                #     => found_primer = {} or {primer}, ambiguous_primers = {}
-                # - "sample_read contains Ns in primer sequence"
-                #     => found_primer = {}, ambiguous_primers = {*}
-                # The found primer (if exists) is removed as that's an exact match
-                ambiguous_primers = (
-                    set(primer_auto.automaton.keys(sample_read, "N", ahocorasick.MATCH_AT_MOST_PREFIX)) - found_primer
-                )
-
-                # increment the read count as necessary
-                if found_primer:
+                # list: (primer_end_position_in_read, primer)
+                primers_found_in_read = list(primer_auto.automaton.iter(sample_read))
```

Comment:
This searches all the primers for a certain scheme in the read window in one go.